### PR TITLE
Surface Arc-Length PE: per-foil chord-wise positional encoding for surface nodes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

On airfoil surfaces, the physically meaningful coordinate is NOT the global (x, y) position but the **arc-length fraction s in [0, 1]** from leading edge to trailing edge. Pressure distributions are smooth functions of arc-length — Cp(s) curves are the standard representation in aerodynamics. By explicitly providing arc-length fraction as a surface-node feature, the model can learn pressure as a 1D function of s, which is the natural basis.

This adds 2 new input channels (sin(2*pi*s), cos(2*pi*s)) for surface nodes only. Volume nodes get zeros. The sin/cos embedding is continuous and handles the LE-TE-LE wraparound naturally.

**This is NOT a repeat of prior arc-length PE experiments.** Prior attempts used arc-length as an additive positional encoding on ALL nodes (confusing volume nodes). This experiment:
- Restricts arc-length to **surface nodes only**
- Normalizes **per-foil** (not globally)
- Uses it as a **physics-motivated feature** (Cp(s) is smooth), not a generic PE
- The sin/cos embedding handles wraparound continuously

**Literature:**
- AeroDiT (arXiv:2412.17394, 2024) — explicit arc-length parameterization of surface pressure, ~10% improvement vs coordinate representations on OOD AoA
- NeuralFoil (arXiv:2503.16323, 2024) — arc-length fraction as primary surface coordinate

### Step 1: Compute Arc-Length Fraction

Add a function to compute per-foil arc-length fraction:

```python
def compute_surface_arclen_fraction(surface_xy, le_xy=None, te_xy=None):
    """
    Compute arc-length fraction for each surface node.
    
    Args:
        surface_xy: [M_s, 2] coordinates of surface nodes for one foil
    
    Returns:
        s_sin_cos: [M_s, 2] sin/cos of 2*pi*s_frac
    """
    # Compute cumulative arc length along surface node sequence
    diffs = torch.diff(surface_xy, dim=0)  # [M_s-1, 2]
    seg_lens = torch.norm(diffs, dim=-1)   # [M_s-1]
    cum_len = torch.cat([torch.zeros(1, device=surface_xy.device), 
                         torch.cumsum(seg_lens, dim=0)])  # [M_s]
    total_len = cum_len[-1].clamp(min=1e-6)
    s_frac = cum_len / total_len  # [M_s] in [0, 1]
    
    # Smooth periodic embedding
    s_sin = torch.sin(2 * math.pi * s_frac)
    s_cos = torch.cos(2 * math.pi * s_frac)
    return torch.stack([s_sin, s_cos], dim=-1)  # [M_s, 2]
```

### Step 2: Integrate into Feature Construction

In the data loading / feature construction pipeline:
1. For each sample, identify surface nodes per foil (use `boundary_id` or surface mask)
2. For each foil's surface nodes, compute arc-length sin/cos features
3. For volume nodes, set these 2 channels to 0.0
4. Append as channels 25-26 of the input feature tensor
5. Update the model's input projection: `nn.Linear(24, n_hidden)` → `nn.Linear(26, n_hidden)`

### Step 3: Handle Node Ordering

**CRITICAL:** The cumulative arc-length computation assumes surface nodes are in order along the airfoil contour. Check whether TandemFoilSet stores surface nodes in arc-length order:
- If YES: use directly
- If NO: sort surface nodes by angle from foil centroid before computing arc-length. This is O(M_s log M_s) per foil per sample — acceptable overhead.

Print the node ordering for a few samples to verify before running training.

### Step 4: New Flag

- `--surface_arclen_pe` (bool, default False) — enables arc-length features, increases input_dim to 26

### Step 5: Training Runs

Run 2 seeds:

```
cd cfd_tandemfoil && python train.py --agent fern --seed 42 \
  --wandb_name "fern/surface-arclen-pe-s42" \
  --wandb_group "surface-arclen-pe" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --surface_arclen_pe
```

Then seed 73 with same flags but `--seed 73`.

### Important Notes

- The per-feature statistics for the 2 new channels should be computed on the training set (they'll be approximately sin/cos distributions — mean near 0, std near 0.7). Include them in the standardization pipeline.
- AoA augmentation (`--aug aoa_perturb --aug_full_dsdf_rot`) rotates the surface coordinates. The arc-length fraction is invariant to rotation (it depends on inter-node distances, not absolute positions), so this is naturally compatible.
- For tandem samples with 2 foils: compute arc-length fraction independently per foil (each foil has its own s in [0,1]).
- The sin/cos embedding avoids the discontinuity at s=0/s=1 (TE) — this is important because the trailing edge is a physically meaningful location where upper/lower surfaces meet.

## Baseline

Current best metrics (PR #2251, 2-seed avg):
- p_in: **11.891** (target: < 11.89)
- p_oodc: **7.561** (target: < 7.56)
- p_tan: **28.118** (target: < 28.12)
- p_re: **6.364** (target: < 6.36)
- W&B baseline runs: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Surface Arc-Length PE: Per-Foil Chord-Wise Positional Encoding

### Surface Pressure MAE (×10³ Pa)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Delta |
|--------|---------|---------|-----------|----------|-------|
| p_in   | 12.755  | 12.346  | **12.550** | 11.891  | +5.5% ❌ |
| p_oodc | 8.259   | 8.030   | **8.145**  | 7.561   | +7.7% ❌ |
| p_tan  | 30.044  | 28.640  | **29.342** | 28.118  | +4.4% ❌ |
| p_re   | 6.446   | 6.460   | **6.453**  | 6.364   | +1.4% ❌ |

### Surface Velocity MAE (2-seed avg)

| Split | Ux | Uy |
|-------|-----|-----|
| in_dist | 2.678 | 0.860 |
| ood_cond | 2.074 | 0.649 |
| tandem | 4.063 | 1.622 |
| ood_re | 1.784 | 0.441 |

### Training Details
- **Epochs:** 144 (s42), 145 (s73) — 180-min timeout
- **Memory:** 46.0-46.3 GB peak
- **Epoch time:** ~75s (no significant overhead from arc-length computation)
- **W&B runs:** `9dbwtqfc` (seed 42), `no0a4utt` (seed 73)
- **W&B group:** `surface-arclen-pe`
- **New flag vs baseline:** `--surface_arclen_pe` only

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/surface-arclen-pe-s42" \
  --wandb_group "surface-arclen-pe" \
  --seed 42 --surface_arclen_pe \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aft_foil_srf_hidden 192 --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### What Happened

**Arc-length PE degraded all metrics**, with p_oodc worst (+7.7%). The degradation is moderate but consistent across all splits and both seeds.

**Why it failed:**

1. **Angle-sort contour ordering is unreliable.** The implementation sorts surface nodes by polar angle from foil centroid (`atan2`). This works for convex shapes but can mis-order nodes near the trailing edge where the angle wraps from +π to -π. For cambered airfoils, TE nodes on the upper and lower surface can have similar angles but very different arc-length positions, creating discontinuous or incorrect arc-length fractions. This pollutes the sin/cos features with noise.

2. **Volume nodes receive zero features — a confusing signal.** The 2 arc-length channels are 0.0 for all volume nodes. The model's attention mechanism sees these zeros as a specific value (not "missing"), creating an artificial distinction between surface and volume nodes. This can interfere with the smooth information flow between surface and volume predictions that the Transolver attention mechanism relies on.

3. **The model already has rich surface-aware features.** The TE coordinate frame (`--te_coord_frame`) provides dx/dy/r offsets from each foil's trailing edge, giving the model chord-wise position information without the ordering ambiguity. The curvature proxy feature and DSDF channels already encode local surface geometry. Adding arc-length on top doesn't provide new information — it's redundant with (and noisier than) existing features.

4. **AoA augmentation rotates the contour.** The `--aug aoa_perturb --aug_full_dsdf_rot` augmentation rotates all coordinates. While arc-length is theoretically rotation-invariant (it depends on inter-node distances), the centroid-angle sorting step is NOT rotation-invariant — rotating the point cloud changes which node gets assigned s=0. This means the same physical location gets different arc-length values under different augmentations, adding training noise.

5. **Seed variance is notable** (seed 42: p_tan=30.04, seed 73: p_tan=28.64), suggesting the noisy arc-length features amplify randomness.

### Suggested Follow-ups

- **Do not pursue arc-length PE** with angle-based sorting. The contour ordering is fundamentally unreliable for unstructured meshes.
- **If arc-length is revisited**, the implementation should use nearest-neighbor contour chaining from a fixed anchor (e.g., TE node), which produces a deterministic, rotation-invariant ordering. However, given that TE coordinate frame already provides chord-wise position, the benefit would be marginal.
- **The high seed variance on p_tan** reinforces that noisy features hurt tandem performance disproportionately.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2278 because ADVISOR: Closing — all 4 metrics regressed (p_in +5.5%, p_oodc +7.7%, p_tan +4.4%, p_re +1.4%).

**Root cause:** Good analysis. The angle-sort contour ordering (atan2) is unreliable near the trailing edge, creating noisy arc-length fractions. Zero features for volume nodes create an artificial surface/volume distinction that confuses attention. Most critically, TE coordinate frame already provides chord-wise position information more robustly and without the ordering ambiguity.

**Key takeaway:** Adding surface-node-only input features with zero-padding for volume nodes disrupts the smooth surface-volume information flow in Transolver attention. Future input features should be defined for ALL nodes (even if the value is less meaningful for some). Added to DO NOT REVISIT.

Moving you to a new direction..
